File: app_flutter/lib/domain/validation.dart
import 'package:app_flutter/domain/domain_errors.dart';
import 'package:app_flutter/domain/result.dart';
import 'package:app_flutter/domain/type_descriptor.dart';

/// Realises: [Feat-10/Validation]
///
/// Generic validation function that evaluates constraints on a map of input values
/// against a list of [FieldDescriptor] descriptors.
///
/// Returns [Result.success] when all inputs satisfy their constraints.
/// Returns [Result.failure] carrying a typed [DomainError] on the first constraint failure.
Result<void> validateFields(Map<String, dynamic> input, List<FieldDescriptor> descriptors) {
  for (final fd in descriptors) {
    final value = input[fd.key];

    // If missing/empty, check required constraint. Otherwise skip validation if not required.
    if (value == null || (value is String && value.trim().isEmpty)) {
      if (fd.required) {
        return Result.failure(SchemaFieldRequiredError(
          fieldName: fd.key,
          schemaName: fd.key,
        ));
      }
      continue;
    }

    final strVal = value.toString();
    if (fd.type == 'int') {
      final parsed = int.tryParse(strVal);
      if (parsed == null) {
        return Result.failure(SchemaFieldTypeError(
          fieldName: fd.key,
          expectedType: 'int',
          actualType: value.runtimeType.toString(),
        ));
      }
      if (fd.minValue != null && parsed < fd.minValue!) {
        return Result.failure(SchemaFieldRangeError(
          fieldName: fd.key,
          value: parsed,
          min: fd.minValue,
          max: fd.maxValue,
        ));
      }
      if (fd.maxValue != null && parsed > fd.maxValue!) {
        return Result.failure(SchemaFieldRangeError(
          fieldName: fd.key,
          value: parsed,
          min: fd.minValue,
          max: fd.maxValue,
        ));
      }
    } else if (fd.type == 'double' || fd.type == 'real') {
      final parsed = double.tryParse(strVal);
      if (parsed == null) {
        return Result.failure(SchemaFieldTypeError(
          fieldName: fd.key,
          expectedType: fd.type,
          actualType: value.runtimeType.toString(),
        ));
      }
      if (fd.minValue != null && parsed < fd.minValue!) {
        return Result.failure(SchemaFieldRangeError(
          fieldName: fd.key,
          value: parsed,
          min: fd.minValue,
          max: fd.maxValue,
        ));
      }
      if (fd.maxValue != null && parsed > fd.maxValue!) {
        return Result.failure(SchemaFieldRangeError(
          fieldName: fd.key,
          value: parsed,
          min: fd.minValue,
          max: fd.maxValue,
        ));
      }
    } else if (fd.type == 'string') {
      if (fd.pattern != null && fd.pattern!.isNotEmpty) {
        final regex = fd.compiledPattern ?? RegExp(fd.pattern!);
        if (!regex.hasMatch(strVal)) {
          return Result.failure(SchemaFieldPatternError(
            fieldName: fd.key,
            value: strVal,
            pattern: fd.pattern!,
          ));
        }
      }
    } else if (fd.type == 'enum') {
      if (fd.enumOptions != null && !fd.enumOptions!.contains(strVal)) {
        return Result.failure(SchemaFieldEnumError(
          fieldName: fd.key,
          value: strVal,
          allowedValues: fd.enumOptions!,
        ));
      }
    }
  }
  return const Result.success(null);
}

File: web_react/src/domain/validation.ts
export interface FieldDescriptor {
  key: string;
  label: string;
  type: string;
  required?: boolean;
  isRequired?: boolean;
  minValue?: number;
  maxValue?: number;
  pattern?: string;
  options?: string[];
  enumOptions?: string[];
}

/**
 * Generic validation function that takes an input map and a list of field descriptors,
 * evaluating constraints (isRequired, minValue/maxValue, pattern regex, options enum).
 */
export function validateFields(input: Record<string, any>, descriptors: FieldDescriptor[]): boolean {
  for (const fd of descriptors) {
    const value = input[fd.key];
    const isRequired = fd.required ?? fd.isRequired ?? false;

    // Check if required. Optional fields that are empty skip constraint validation.
    if (value === undefined || value === null || (typeof value === 'string' && value.trim() === '')) {
      if (isRequired) {
        return false;
      }
      continue;
    }

    const strVal = String(value);
    const type = fd.type;

    if (type === 'int') {
      const parsed = parseInt(strVal, 10);
      if (isNaN(parsed) || parsed.toString() !== strVal) return false;
      if (fd.minValue !== undefined && fd.minValue !== null && parsed < fd.minValue) return false;
      if (fd.maxValue !== undefined && fd.maxValue !== null && parsed > fd.maxValue) return false;
    } else if (type === 'double' || type === 'real') {
      const parsed = parseFloat(strVal);
      if (isNaN(parsed)) return false;
      if (fd.minValue !== undefined && fd.minValue !== null && parsed < fd.minValue) return false;
      if (fd.maxValue !== undefined && fd.maxValue !== null && parsed > fd.maxValue) return false;
    } else if (type === 'string') {
      if (fd.pattern) {
        const regex = new RegExp(fd.pattern);
        if (!regex.test(strVal)) return false;
      }
    } else if (type === 'enum') {
      const options = fd.options ?? fd.enumOptions;
      if (options && !options.includes(strVal)) {
        return false;
      }
    }
  }
  return true;
}
